Marlin/ultralcd.cpp

```diff
@@ -2847,101 +2853,94 @@ void lcd_update() {
-              case LCDVIEW_REDRAW_NOW:        // set above, or by a handler through LCDVIEW_CALL_REDRAW_NEXT
-              case LCDVIEW_NONE:
-                break;
-            } // switch
-          }
-        #if ENABLED(ULTIPANEL)
-          #define CURRENTSCREEN() (*currentScreen)(), lcd_clicked = false
-        #else
-          #define CURRENTSCREEN() lcd_status_screen()
-        #endif
-
-        #if ENABLED(DOGLCD)  // Changes due to different driver architecture of the DOGM display
-          if (!drawing_screen) {
-            u8g.firstPage();
-            drawing_screen = 1;
-          }
-          lcd_setFont(FONT_MENU);
-          CURRENTSCREEN();
-          if (drawing_screen && (drawing_screen = u8g.nextPage())) return;
-        #else
-          CURRENTSCREEN();
-        #endif
-      }
-
-      #if ENABLED(ULTIPANEL)
-
-        // Return to Status Screen after a timeout
-        if (currentScreen == lcd_status_screen || defer_return_to_status)
-          return_to_status_ms = ms + LCD_TIMEOUT_TO_STATUS;
-        else if (ELAPSED(ms, return_to_status_ms))
-          lcd_return_to_status();
-
-      #endif // ULTIPANEL
+    millis_t bbr = planner.block_buffer_runtime();
 
+    #if ENABLED(DOGLCD)
+      if ((lcdDrawUpdate || drawing_screen) && (!bbr || (bbr > max_display_update_time*2000)))
+    #else
+      if (lcdDrawUpdate && (!bbr || (bbr > max_display_update_time*2000)))
+    #endif
+    {
       #if ENABLED(DOGLCD)
         if (!drawing_screen)
       #endif
         {
           switch (lcdDrawUpdate) {
-            case LCDVIEW_CLEAR_CALL_REDRAW:
-              lcd_implementation_clear();
-            case LCDVIEW_CALL_REDRAW_NEXT:
-              lcdDrawUpdate = LCDVIEW_REDRAW_NOW;
-              break;
-            case LCDVIEW_REDRAW_NOW:
+            case LCDVIEW_CALL_NO_REDRAW:
               lcdDrawUpdate = LCDVIEW_NONE;
               break;
+            case LCDVIEW_CLEAR_CALL_REDRAW: // set by handlers, then altered after (rarely occurs here)
+            case LCDVIEW_CALL_REDRAW_NEXT:  // set by handlers, then altered after (never occurs here?)
+              lcdDrawUpdate = LCDVIEW_REDRAW_NOW;
+            case LCDVIEW_REDRAW_NOW:        // set above, or by a handler through LCDVIEW_CALL_REDRAW_NEXT
             case LCDVIEW_NONE:
               break;
           } // switch
         }
-    } // LCD_HANDLER_CONDITION
+      #if ENABLED(ULTIPANEL)
+        #define CURRENTSCREEN() (*currentScreen)(), lcd_clicked = false
+      #else
+        #define CURRENTSCREEN() lcd_status_screen()
+      #endif
+
+      #if ENABLED(DOGLCD)  // Changes due to different driver architecture of the DOGM display
+        if (!drawing_screen) {
+          u8g.firstPage();
+          drawing_screen = 1;
+        }
+        lcd_setFont(FONT_MENU);
+        CURRENTSCREEN();
+        if (drawing_screen && (drawing_screen = u8g.nextPage())) {
+          max_display_update_time = max(max_display_update_time, millis() - ms);
```

Comment:
1 ) You got it.

2) We are not just skipping "return to status screen" and "updating the state" here.
The trick is to set `drawing_screen` when the screen content changes. The remaining stripes of a outdated screen are not drawn anymore - but the new screen will begin to draw immediately. Normally this skips 3 stripes. When 'turning' in the menu you will see the difference.
An important difference to before is line 2763-2767. We do not wait anymore until the 100ms are over while we are drawing. That's the reason for the `lcd_status_update_delay = 3 + 9;`.  9 used to be 1 second. Now 9 is 1 second minus the up to 3 stripes we don't wait for.
Returning gives the chance to refill the buffer a bit before trying to draw the next part.

Look at the times. ~Once per second a burst of 4 stripe-updates. Done after ~52ms - that's very cloth to the theoretical minimum.
```
13:46:02.209 : ;19*0
13:46:02.226 : ;19*0
13:46:03.226 : ;18*0
13:46:03.246 : ;18*0
13:46:03.262 : ;18*0
13:46:03.278 : ;18*0
13:46:04.278 : ;17*0
13:46:04.294 : ;17*0
13:46:04.310 : ;19*0
13:46:04.327 : ;19*0
13:46:05.326 : ;18*0
```